src/components/utils/useListNavigation/useListNavigation.ts

```diff
@@ -0,0 +1,179 @@
+import {RefObject, useCallback, useEffect, useLayoutEffect, useState} from 'react';
+
+import {moveBack} from './moveBack';
+import {moveForward} from './moveForward';
+
+export type UseListNavigationProps<ItemType, AnchorType> = {
+    items: ItemType[];
+    onSelect: (item: ItemType, event: KeyboardEvent) => void;
+    skip?: (item: ItemType) => boolean;
+    pageSize?: number;
+    processHomeKey?: boolean;
+    processEndKey?: boolean;
+    disabled?: boolean;
+    initialValue?: number;
+    anchorRef?: RefObject<AnchorType>;
+    onAnchorKeyDown?: (event: KeyboardEvent) => void | boolean;
+};
+
+export function useListNavigation<ItemType, AnchorType extends HTMLElement>({
+    items,
+    onSelect,
+    skip,
+    pageSize,
+    processHomeKey = true,
+    processEndKey = true,
+    anchorRef,
+    disabled = false,
+    initialValue = -1,
+    onAnchorKeyDown,
+}: UseListNavigationProps<ItemType, AnchorType>) {
+    const [activeItemIndex, setActiveItemIndex] = useState<number>(initialValue);
+
+    const reset = useCallback(() => {
+        setActiveItemIndex(initialValue);
+    }, [initialValue]);
+
+    useEffect(() => {
+        if (items) {
+            reset();
+        }
+    }, [items, reset]);
+
+    useLayoutEffect(() => {
+        if (disabled) {
+            return;
+        }
+
+        const anchor = anchorRef?.current;
+        if (!anchor) {
+            return;
+        }
+
+        const handleKeyDown = (event: KeyboardEvent) => {
+            const shouldProcessKeydown = onAnchorKeyDown?.(event);
+            if (shouldProcessKeydown === false) {
+                return;
+            }
+
+            switch (event.key) {
+                case 'ArrowDown':
+                case 'ArrowRight': {
+                    event.preventDefault();
+
+                    // Go 1 step forward
+                    const newActiveItemIndex = moveForward(items, activeItemIndex, 1, skip);
+                    setActiveItemIndex(newActiveItemIndex);
```

Comment:
I believe it is better to modify state via callback, you can avoid `activeItemIndex` in deps array

```js
setActiveItemIndex((activeItemIndex) => moveForward(items, activeItemIndex, 1, skip))
```